rpc/dialer.go

```diff
@@ -240,6 +241,145 @@ func DialDirectGRPC(ctx context.Context, address string, logger utils.ZapCompati
 	return dialInner(ctx, address, logger, dOpts)
 }
 
+func socksProxyDialContext(ctx context.Context, network, proxyAddr, addr string) (net.Conn, error) {
+	dialer, err := proxy.SOCKS5(network, proxyAddr, nil, proxy.Direct)
+	if err != nil {
+		return nil, fmt.Errorf("error creating SOCKS proxy dialer to address %q from environment: %w",
+			proxyAddr, err)
+	}
+	conn, err := dialer.(proxy.ContextDialer).DialContext(ctx, network, addr)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+// SocksProxyFallbackDialContext will return nil if SocksProxyEnvVar is not set or if trying to connect to a local address,
+// which will allow dialers to use the default DialContext.
+// If SocksProxyEnvVar is set, it will prioritize a connection made without a proxy but will fall back to a SOCKS proxy connection.
+func SocksProxyFallbackDialContext(
+	ctx context.Context,
+	network, addr string,
+	logger utils.ZapCompatibleLogger,
+) func(ctx context.Context, network, addr string) (net.Conn, error) {
+	// Use SOCKS proxy from environment as gRPC proxy dialer. Do not use SOCKS proxy if trying to connect to a local address.
+	localAddr := strings.HasPrefix(addr, "[::]") || strings.HasPrefix(addr, "localhost") || strings.HasPrefix(addr, "unix")
+	proxyAddr := os.Getenv(SocksProxyEnvVar)
+	if localAddr || proxyAddr == "" {
+		// return nil in these cases so that the default dialer gets used instead.
+		return nil
+	}
+
+	return func(ctx context.Context, network, addr string) (net.Conn, error) {
+		// if ONLY_SOCKS_PROXY specified, no need for a parallel dial - only dial through
+		// the SOCKS proxy directly.
+		if os.Getenv(OnlySocksProxyEnvVar) != "" {
+			logger.Infow("Both SOCKS_PROXY and ONLY_SOCKS_PROXY specified, only SOCKS proxy will be used for outgoing connection")
+			conn, err := socksProxyDialContext(ctx, network, proxyAddr, addr)
+			if err == nil {
+				logger.Infow("connected with SOCKS proxy")
+			}
+			return conn, err
+		}
+
+		type dialResult struct {
```

Comment:
reference https://go.dev/src/net/dial.go#L585

Reply:
Could you leave this reference in-line in the code, too? Just in case we come back to it later.